File: flutter_campus_talk/lib/models/comment.dart
import 'user.dart';

class Comment {
  final int id;
  final User author;
  final int postId;
  final int? parentCommentId;
  final String content;
  final String createdAt;
  final List<Comment>? replies;

  Comment({
    required this.id,
    required this.author,
    required this.postId,
    this.parentCommentId,
    required this.content,
    required this.createdAt,
    this.replies,
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
      id: json['id'], 
      author: User.fromJson(json['author']), 
      postId: json['postId'], 
      parentCommentId: json['parent_comment_id'],
      content: json['content'], 
      createdAt: json['created_at'],
      replies: (json['replies'] as List<dynamic>?)
        ?.map((replyJson) => Comment.fromJson(replyJson))
        .toList(),
    );
  }
}

File: laravel_campus_talk/app/Models/Comment.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Comment extends Model
{
    use HasFactory;

    protected $fillable = [
        'user_id',
        'post_id',
        'parent_comment_id',
        'content',
    ];

    // Relasi: Comment dimiliki oleh satu User (penulis komentar)
    public function author()
    {
        return $this->belongsTo(User::class, 'user_id');
    }

    // Relasi: Comment dimiliki oleh satu Post
    public function post()
    {
        return $this->belongsTo(Post::class);
    }

    // Relasi: Comment bisa memiliki komentar induk (parent)
    public function parent()
    {
        return $this->belongsTo(Comment::class, 'parent_comment_id');
    }

    // Relasi: Comment bisa memiliki banyak balasan (replies)
    public function replies()
    {
        return $this->hasMany(Comment::class, 'parent_comment_id')->with('author', 'replies'); // Dengan eager loading rekursif
    }
}